chalk-derive/src/lib.rs

```diff
@@ -30,11 +30,14 @@ pub fn derive_fold(item: TokenStream) -> TokenStream {
             impl #impl_generics Fold < #arg, #arg > for #type_name #ty_generics #where_clause_ref {
                 type Result = Self;
 
-                fn fold_with(
+                fn fold_with<'i>(
                     &self,
-                    folder: &mut dyn Folder < #arg, #arg >,
+                    folder: &mut dyn Folder < 'i, #arg, #arg >,
                     binders: usize,
-                ) -> ::chalk_engine::fallible::Fallible<Self::Result> {
+                ) -> ::chalk_engine::fallible::Fallible<Self::Result>
+                where
```

Comment:
Nit: this diff is probably not needed, but I guess it doesn't hurt -- in particular `#arg` will normally be a type name like `ChalkIr`, so this is "clearly true"